source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +90,61 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    applySubstitutionFormatter(config_->descriptors(),
+                               filter_callbacks_->connection().streamInfo());
+  }
+
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                  StreamInfo::StreamInfo& stream_info) {
+
+    std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+    for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+      RateLimit::Descriptor new_descriptor;
+      for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+        std::string value = descriptorEntry.value_;
+        if (config_->dynamicDownstreamIp() &&
+            absl::StrContains(descriptorEntry.value_, "DOWNSTREAM_REMOTE_ADDRESS_WITHOUT_PORT")) {
```

Comment:
You shouldn't need to do any of this. The idea is to pre-create all of the formatters when you read the config, and then just apply them directly during proxying.

Reply:
Just to clarify: what we want to do is allow the user to add substitution strings to the existing `RateLimitDescriptor.Entry.value` in their filter config, and the only change is that we run `substitutionFormattedString` on those now?

Reply:
Yes, sorry for not being more clear.

The idea is that we should parse the entry value into a substitution formatter on configuration load. The common case will be a string -> string mapping which will have negligible perf impact. Otherwise, we can use the formatter when populating the real entry that gets sent to the server given the current stream info in the filter chain. Let me know if you need more code pointers.

Reply:
@mattklein123 the stream_info used to scrape the information isn't initialized in the config. I do understand that we just want to generally apply the substitution formatter to these descriptors now though.

Reply:
Yes, stream info is supplied at request time, not at config load time, into the pre created formatter.